### PR TITLE
Fix incorrect authentication latency metric

### PR DESCRIPTION
**Pull request by marseel**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Authentication latency metric was incorrectly calculated. It took into account request duration. 

#### Does this PR introduce a user-facing change?

```release-note
Fixed authentication_duration_seconds metric. Previously it included whole apiserver request duration. 
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go [hunk 1]
```diff
@@ -47,7 +47,7 @@ func WithAuthentication(handler http.Handler, auth authenticator.Request, failed
 			req = req.WithContext(authenticator.WithAudiences(req.Context(), apiAuds))
 		}
 		resp, ok, err := auth.AuthenticateRequest(req)
-		defer recordAuthMetrics(req.Context(), resp, ok, err, apiAuds, authenticationStart)
+		defer recordAuthMetrics(req.Context(), resp, ok, err, apiAuds, authenticationStart, time.Now())
```
Comment (mborsz) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go:50]: Why do we need to defer here?

Comment (marseel) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go:50]: I don't think we need it. I could remove defer too.

Comment (marseel) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: Done.

**Comment by wojtek-t** [comment]
/hold

@marseel - please add a release note
I would say that we should probably cherrypick this as the metrics were broken previously (@liggitt - for thoughts)

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go [hunk 2]
```diff
@@ -47,7 +47,7 @@ func WithAuthentication(handler http.Handler, auth authenticator.Request, failed
 			req = req.WithContext(authenticator.WithAudiences(req.Context(), apiAuds))
 		}
 		resp, ok, err := auth.AuthenticateRequest(req)
-		defer recordAuthMetrics(req.Context(), resp, ok, err, apiAuds, authenticationStart)
+		recordAuthMetrics(req.Context(), resp, ok, err, apiAuds, authenticationStart)
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go:50]: /hold

This breaks recording of the authentication audience error case on line 60. The fact that unit tests are green mean we have a test gap that should be closed as well.

Once this is correct, I would support merge and backport

Comment (marseel) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: I've added tests. Also, this error on line 60 was not working before. When you do `defer f(err)` err is evaluated at the time of defer, not later.

**Comment by liggitt** [comment]
/lgtm
/approve
/hold cancel
/priority important-soon
/milestone v1.21
/kind regression

fixes regression in 1.18+ that breaks the ability to determine how long the authentication phase of the request is taking

**Comment by ialidzhikov** [comment]
@marseel , can you file the corresponding cherry-pick PRs?
Ref https://github.com/kubernetes/community/blob/master/contributors/devel/sig-release/cherry-picks.md

**Pull request merged**